pkgs/by-name/ol/ollama/package.nix

```diff
@@ -118,16 +118,12 @@ let
     appleFrameworks.MetalPerformanceShaders
   ];
 
-  runtimeLibs = lib.optionals enableRocm [
-    rocmPath
-  ] ++ lib.optionals enableCuda [
-    linuxPackages.nvidia_x11
-  ];
-  wrapperOptions = builtins.concatStringsSep " " ([
-    "--suffix LD_LIBRARY_PATH : '/run/opengl-driver/lib:${lib.makeLibraryPath runtimeLibs}'"
+  wrapperOptions = [
+    "--suffix LD_LIBRARY_PATH : '/run/opengl-driver/lib'"
```

Comment:
1. `/run/opengl-driver/lib` should be `${addDriverRunpath.driverLink}/lib` to simplify future migrations
1. Does `ollama` actually need `LD_LIBRARY_PATH`? I see it's a native program, is it not enough tthat `addDriverRunpath` patches dt `DT_RUNPATH`s?

Reply:
> Does `ollama` actually need `LD_LIBRARY_PATH`?

I don't really know. I normally would have tested that locally, but am unable to do so at the moment, so I was more cautious and decided to leave it in case it was necessary.

Since you seem to think it's unnecessary, I just pushed some code removing the wrapper. Hopefully rocm and cuda both still work right, though I will need others to test both to see if they use the gpu.

@mrdev023 Could you test the new version to see if cuda still works right?

Reply:
I'm prepare my conf to test it

Reply:
I see this

```
llama3:latest	365c0bd3c000	5.4 GB	100% GPU 	4 minutes from now	
```

But is slow and use CPU and not GPU. I'll try with switch / boot instead of test

Reply:
I confirm it's a regretion, when i rollback and return back to the global settings 

```nix
services.ollama = {
      enable = true;

      acceleration = if nvidiaEnabled then "cuda" else null;
    };
```

It works well

Reply:
I don't know if yesterday i'll try to use chat to check the speed.

And i don't know if is a configuration regression or ollama regression because ollama ps display it use GPU at 100% but it's wrong

Reply:
Currently in nixos-unstable, i'm using ollama version 0.1.45

Reply:
Ok, I created a reproducer: https://gist.github.com/SomeoneSerge/309d4e26fdf0af328968cfcfac33347e

Basically, `ollama serve` does successfully load `libcuda.so` using `DT_RUNPATH`: https://gist.github.com/SomeoneSerge/309d4e26fdf0af328968cfcfac33347e#file-run-ollama-log-L111

However, it also generates another executable on the fly (maybe during `ollama run`), and that executable is incorrect and fails to load `libcuda.so`:
```
        61:	 search path=/nix/store/hzh8w7jkkqf6wlngvy0rb26bxjzphcpz-cuda-toolkit/lib/glibc-hwcaps/x86-64-v3:...:/nix/store/avkdrmgnn4c0b4sqnlgbn0w464cavh33-cuda_cudart-12.2.140/lib
                		(RUNPATH from file /build/ollama1724294767/runners/cuda_v12/ollama_llama_server)
```

Reply:
@abysssol let's temporarily recover `LD_LIBRARY_PATH` in the wrapper and add a comment that it's to affect the "runner" executables created by ollama on the fly (or whatever that means, I'm not familiar with ollama)

Reply:
Btw I'd be grateful if you could help me wrap this into a proper test. Right now I'm confused about ollama's api, with the driver issue addressed it still fails with:

```
llama_new_context_with_model: graph splits = 2
⠋ INFO [main] model loaded | tid="140737352847360" timestamp=1719755151
time=2024-06-30T13:45:51.840Z level=INFO source=server.go:599 msg="llama runner started in 0.91 seconds"
[GIN] 2024/06/30 - 13:45:51 | 200 |  1.071189722s |       127.0.0.1 | POST     "/api/generate"
[GIN] 2024/06/30 - 13:45:51 | 400 |       32.77µs |       127.0.0.1 | POST     "/api/chat"
Traceback (most recent call last):
  File "/build/.attr-1h44fmdnc6dwaj0f8gdbv6z0mrdzkgx9p7zjc2girsm2j4cbi4g2", line 43, in <module>
    r.raise_for_status()
  File "/nix/store/srfswwz4ma22cqjmydjb36l40ki0rmak-python3-3.11.9-env/lib/python3.11/site-packages/requests/models.py", line 1024, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: http://127.0.0.1:11434/api/chat
```

...at https://gist.github.com/SomeoneSerge/8207f77c78e92e787d6d1114b01ae51c#file-llama-2-7b-nix-L108-L117

Reply:
Maybe they're not really generated on the fly: https://github.com/ollama/ollama/blob/c1218199cfe82eda35f5e4a8031eee28f01ebf75/llm/payload.go#L162. Maybe we could patchelf them during the build if we did it early enough

Reply:
> Maybe they're not really generated on the fly

No, they're not. They are compiled and compressed by `go generate` from ollama's custom c++ code which wraps llama-cpp's code/api, then they're embedded into the ollama binary. Ollama merely extracts them and runs them.

Ollama implements its gpu support by building multiple of these binaries, for example one that supports cuda, another that supports rocm, another that supports cpus that support a certain feature/instruction, and another fallback cpu build that is expected to work on almost all x86 cpus. At runtime, it tries running the binaries with gpu support, and if they fail it falls back to the binaries with cpu support.

This is what makes using the version of llama-cpp packaged in nix so difficult. Ollama has significant amounts of custom code that would need to be patched/spliced into llama-cpp's source, and it might require some modification to make it work right in a different environment (if ollama's custom files have to be moved, paths that refer to them would have to be updated, etc). Ollama also maintains patches for llama-cpp that are specific to particular versions of llama-cpp, so any update of llama-cpp's source would require those patches to be inspected, maybe removed or updated, which is one reason why ollama doesn't use the latest version of llama-cpp.

This applies to #323672: "reus[ing] the prebuilt llama-cpp from Nixpkgs" isn't really possible because ollama doen't use stock llama-cpp in the first place. However, I do think separating the `go generate` step into a separate derivation is still a good idea, and shouldn't be too difficult.

All of this is why I still haven't tried to replace the vendored llama-cpp with nix's llama-cpp package: it's been too intimidating, and seemed to require far more effort than I wanted to put in, especially when ollama works fine without doing so. In the long term, it might make sense to integrate into nix's llama-cpp package, but I would expect significant changes to the llama-cpp package to be required, and using the same version of llama-cpp's source as ollama expects will probably always be necessary.

> Maybe we could patchelf them during the build if we did it early enough

I think the script run by `go generate` would have to be patched to achieve that, though I'm not sure where exactly in the script that would have to go.

If I can find the right place, what would I have to put there? Should I just use patchelf to add `libcuda.so` to the binary's runpath, or is there some way to use whatever autoAddDriverRunpath does on arbitrary binaries?

> Btw I'd be grateful if you could help me wrap this into a proper test

I can't see anything obviously wrong at the moment; I'll look closer when I get the chance.

I never realized that gpu support was possible in the nixos test vms. Good to see that it actually is, though I wonder how that interacts with local test execution. If the local machine doesn't support cuda, do the tests that require cuda simply not run?

Reply:
> If I can find the right place, what would I have to put there? Should I just use patchelf to add libcuda.so to the binary's runpath, or is there some way to use whatever autoAddDriverRunpath does on arbitrary binaries?

Something like `addDriverRunpath cuda_v12_runner.so` (warning: a purely made up name) after `go generate`

> This applies to https://github.com/NixOS/nixpkgs/issues/323672: "reus[ing] the prebuilt llama-cpp from Nixpkgs" isn't really possible because ollama doen't use stock llama-cpp in the first place

I think this might be feasible after all, but let's not get ahead of ourselves.


For now, let's just recover the `wrapProgram` logic and rely on `LD_LIBRARY_PATH` to handle the embedded runners

Reply:
> If the local machine doesn't support cuda, do the tests that require cuda simply not run?

Nix would reject the build with a message "error: a 'x86_64-linux' with features {cuda} is required to build ...drv, ..."

> I never realized that gpu support was possible in the nixos test vms.

This one doesn't involve VMs, just simple derivations built in the sandbox. This is experimental and a recent introduction: https://github.com/NixOS/nixpkgs/pull/256230/commits/79a7186f1ce8d94b0c136a7cc7c3e2e31facc794

Reply:
The `${addDriverRunpath.driverLink}/lib` comment still applies:)

Reply:
As well as

> add a comment that it's to affect the "runner" executables